lib/firmware/fsm_msg_nano.h

```diff
@@ -0,0 +1,383 @@
+             address_n[2] & 0x7ffffff);
+    return true;
+}
+
+void fsm_msgNanoGetAddress(NanoGetAddress *msg)
+{
+    RESP_INIT(NanoAddress);
+
+    CHECK_INITIALIZED
+
+    CHECK_PIN
+
+    const char *coin_name = msg->has_coin_name ? msg->coin_name : "Nano";
+    const CoinType *coin = fsm_getCoin(true, coin_name);
+    if (!coin) return;
+    HDNode *node = fsm_getDerivedNode(coin->curve_name, msg->address_n, msg->address_n_count, NULL);
+    if (!node) return;
+    hdnode_fill_public_key(node);
+
+    char address[MAX_NANO_ADDR_SIZE];
+    if (!nano_get_address(
+        &node->public_key[1],
+        coin->nanoaddr_prefix, strlen(coin->nanoaddr_prefix),
+        address, sizeof(address))) {
+        fsm_sendFailure(FailureType_Failure_Other, _("Can't encode address"));
+        layoutHome();
+        return;
+    }
+
+    if (msg->has_show_display && msg->show_display) {
+        char node_str[NODE_STRING_LENGTH];
+        if (!nano_bip32_to_string(node_str, sizeof(node_str), coin, msg->address_n,
+                                  msg->address_n_count) &&
+            !bip32_path_to_string(node_str, sizeof(node_str),
+                                  msg->address_n, msg->address_n_count)) {
+            memset(node_str, 0, sizeof(node_str));
+        }
+
+        bool mismatch = nano_path_mismatched(coin, msg->address_n, msg->address_n_count);
+
+        if (mismatch) {
+            if (!confirm(ButtonRequestType_ButtonRequest_Other, "WARNING", "Wrong address path for selected coin. Continue at your own risk!")) {
+                fsm_sendFailure(FailureType_Failure_ActionCancelled, NULL);
+                layoutHome();
+                return;
+            }
+        }
+
+        if(!confirm_nano_address(node_str, address)) {
+            fsm_sendFailure(FailureType_Failure_ActionCancelled, "Show address cancelled");
+            layoutHome();
+            return;
+        }
+    }
+
+    strlcpy(resp->address, address, sizeof(resp->address));
+    msg_write(MessageType_MessageType_NanoAddress, resp);
+    layoutHome();
+}
+
+void nano_hash_block_data(const ed25519_public_key account_pk,
+                          const uint8_t parent_hash[32],
+                          const uint8_t link[32],
+                          const ed25519_public_key representative_pk,
+                          const uint8_t balance[16],
+                          uint8_t out_hash[32])
+{
+    blake2b_state ctx;
+    blake2b_Init(&ctx, 32);
+
+    blake2b_Update(&ctx, NANO_BLOCK_HASH_PREAMBLE, sizeof(NANO_BLOCK_HASH_PREAMBLE));
+    blake2b_Update(&ctx, account_pk, sizeof(ed25519_public_key));
+    blake2b_Update(&ctx, parent_hash, 32);
+    blake2b_Update(&ctx, representative_pk, sizeof(ed25519_public_key));
+    blake2b_Update(&ctx, balance, 16);
+    blake2b_Update(&ctx, link, 32);
+
+    blake2b_Final(&ctx, out_hash, 32);
+}
+
+void nano_truncate_address(const CoinType *coin, char *str) {
```

Comment:
I'd move these helpers to a new file: `lib/firmware/nano.c`, which makes them unit-testable.

Reply:
Done and added some unit-tests as well.